New quickstart template for GlassFish AS on SUSE

CalCof opened the pull request:
This quickstart template creates a small load balanced GlassFish application server (v3 or v4) cluster on SUSE linux (SLES or OpenSUSE). It has been very useful for quickly spinning up a GlassFIsh dev/test environments.

azurecla commented:
Hi __@CalCof__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.azure.com.
    </span>

TTYL, AZPRBOT;

azurecla commented:
__@CalCof__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, AZPRBOT;

bmoore-msft commented on glassfish-on-suse/README.md:
```diff
@@ -0,0 +1,74 @@
+15. sshPublicKey: The public key used to secure SSH access with each VM 
+
+
+## Deploy Template
```
use instructions for scripts in the repo (see readme.md)

bmoore-msft commented on glassfish-on-suse/azuredeploy.json:
```diff
@@ -0,0 +1,539 @@
+  "$schema": "http://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
+  "contentVersion": "1.0.0.0",
+  "parameters": {
+    "artifactsLocation": {
```
see bp checklist for using artifacts in samples

bmoore-msft commented on glassfish-on-suse/azuredeploy.json:
```diff
@@ -0,0 +1,539 @@
+    "numberOfInstances": {
+      "type": "int",
+      "minValue": 1,
+      "allowedValues": [ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 ],
```
use min/max

bmoore-msft commented on glassfish-on-suse/azuredeploy.json:
```diff
@@ -0,0 +1,539 @@
+        "description": "Number of VMs"
+      }
+    },
+    "storageAccountName": {
```
generate this in a variable

bmoore-msft commented on glassfish-on-suse/azuredeploy.json:
```diff
@@ -0,0 +1,539 @@
+        "description": "Version of GlassFish to deploy"
+      }
+    },
+    "vmPrefix": {
```
move to variable

bmoore-msft commented on glassfish-on-suse/azuredeploy.json:
```diff
@@ -0,0 +1,539 @@
+        "description": "Name of storage account"
+      }
+    },
+    "publicIPAddressName": {
```
move to variable

bmoore-msft commented on glassfish-on-suse/azuredeploy.json:
```diff
@@ -0,0 +1,539 @@
+        "description": "DNS for Load Balancer IP"
+      }
+    },
+    "vnetName": {
```
move to variable

bmoore-msft commented on glassfish-on-suse/azuredeploy.json:
```diff
@@ -0,0 +1,539 @@
+        "description": "VNET name"
+      }
+    },
+    "lbName": {
```
move to variable

bmoore-msft commented on glassfish-on-suse/azuredeploy.json:
```diff
@@ -0,0 +1,539 @@
+    "vmSize": {
+      "type": "string",
+      "defaultValue": "Standard_DS2_v2",
+      "allowedValues": [
```
remove allowed values, many are not supported in other clouds

bmoore-msft commented on glassfish-on-suse/azuredeploy.json:
```diff
@@ -0,0 +1,539 @@
+  },
+  "variables": {
+    "location": "[resourceGroup().location]",
+    "apiVersionStorage": "2015-06-15",
```
use literal values for apiVersion so schema validation can be done

CalCof replied:
Do you mean I should avoid listing apiVersions as variables? It seems much more flexible to only list the values once.

bmoore-msft commented on glassfish-on-suse/azuredeploy.json:
```diff
@@ -0,0 +1,539 @@
+    "lbProbeName": "[concat(variables('namePrefix'), 'tcpProbe')]",
+    "lbProbeID": "[concat(variables('lbID'),'/probes/', variables('lbProbeName'))]",
+    "vnetID": "[resourceId('Microsoft.Network/virtualNetworks',parameters('vnetName'))]",
+    "subnetRef": "[concat(variables('vnetID'),'/subnets/', variables ('subnetName'))]",
```
simplify to: resourceId('Microsoft.Network/virtualNetworks/subnets', variables('vnetName'), variables('subnetname')

bmoore-msft commented on glassfish-on-suse/azuredeploy.json:
see bp #9

bmoore-msft commented on glassfish-on-suse/azuredeploy.json:
see BP #9

bmoore-msft commented on glassfish-on-suse/azuredeploy.json:
move to child resource

CalCof commented on glassfish-on-suse/azuredeploy.json:
Out of curiosity, why should this extension be implemented as a child resource?

CalCof commented:
I believe I have now accommodated all the changes you requested, and I have obviously confirmed that the template still works as expected, but just let me know if you require any additional updates.

bmoore-msft commented on glassfish-on-suse/README.md:
```diff
@@ -0,0 +1,71 @@
+You'll need to generate a pair of SSH keys in order to provision this template.<br/>
+If you are using a Windows computer, then you can download puttygen.exe to create a valid key pair. Alternativley, on a Linux or Mac, you can just use the ssh-keygen command.
+
+### azuredeploy.Parameters.json File Explained
```
some of these have been moved to var, so update this section

CalCof replied:
I have incorporated your comment, but as it was placed on the line above the parameters, it is not automatically hidden

bmoore-msft commented on glassfish-on-suse/README.md:
```diff
@@ -0,0 +1,71 @@
+12. sshPublicKey: The public key used to secure SSH access with each VM 
+
+
+## Deploy Template
```
Update this section to use the scripts in the root of the repo: https://github.com/Azure/azure-quickstart-templates#deploying-samples

bmoore-msft commented on glassfish-on-suse/azuredeploy.json:
```diff
@@ -0,0 +1,462 @@
+    "lbPoolID": "[concat(variables('lbID'),'/backendAddressPools/', variables('lbPoolName'))]",
+    "lbProbeName": "[concat(variables('namePrefix'), 'tcp-probe')]",
+    "lbProbeID": "[concat(variables('lbID'),'/probes/', variables('lbProbeName'))]",
+    "vnetID": "[resourceId('Microsoft.Network/virtualNetworks', variables('vnetName'))]",
```
remove this - it's not used

bmoore-msft commented on glassfish-on-suse/azuredeploy.json:
see BP checklist re: handling uris/endpoints

bmoore-msft commented on glassfish-on-suse/azuredeploy.json:
see BP checklist for handling uris/endpoints

bmoore-msft commented on glassfish-on-suse/azuredeploy.parameters.json:
```diff
@@ -0,0 +1,36 @@
+{
```
remove parameters that have default values

bmoore-msft replied:
When a parameter has a defaultValue declared in the template (e.g. _artifactsLocation, vmsize, osversion - they aren't needed here...

bmoore-msft commented on glassfish-on-suse/azuredeploy.parameters.json:
```diff
@@ -0,0 +1,36 @@
+    "osImage": {
+      "value": "opensuse"
+    },
+    "adminUsername": {
```
GEN-UNIQUE

CalCof replied:
I have done this, but the comment has not been removed as it was placed on the line above the value

bmoore-msft commented on glassfish-on-suse/azuredeploy.parameters.json:
```diff
@@ -0,0 +1,36 @@
+    "adminUsername": {
+      "value": "change-me"
+    },
+    "glassfishAdminPassword": {
```
GEN-PASSWORD

CalCof commented on glassfish-on-suse/azuredeploy.json:
Can you please be more specific regarding what you would like me to change, because I cannot seem to find anything that suggests this format is incorrect and this appears to be the format used in the other quickstart directories.

CalCof commented on glassfish-on-suse/azuredeploy.json:
Can you please be more specific regarding what you would like me to change, because I cannot seem to find anything that suggests this format is incorrect and this appears to be the format used in the other quickstart directories.

bmoore-msft commented on glassfish-on-suse/README.md:
````diff
@@ -45,14 +45,14 @@ There are several ways in which you can deploy this template:
+CLI
   ```
-.\Deploy-AzureResourceGroup.ps1 -ResourceGroupLocation <Location> -ArtifactStagingDirectory 'glassfish-on-suse' 
+azure-group-deploy.sh -a 'glassfish-on-suse' -l <Location>
````
you need to -u switch for the nested templates

bmoore-msft commented on glassfish-on-suse/README.md:
````diff
@@ -45,14 +45,14 @@ There are several ways in which you can deploy this template:
+PowerShell
   ```
-azure-group-deploy.sh -a 'glassfish-on-suse' -l <Location>
+.\Deploy-AzureResourceGroup.ps1 -ResourceGroupLocation <Location> -ArtifactStagingDirectory 'glassfish-on-suse' 
````
here, the -UploadArtifacts switch

bmoore-msft commented on glassfish-on-suse/azuredeploy.json:
See #9 in https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md

bmoore-msft commented on glassfish-on-suse/azuredeploy.json:
See #9 in https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md

CalCof commented:
I believe I have now addressed your remaining comments, but please let me know if there are any additional changes that need made.

I noticed that the template validation is now failing after I changed the template URL to reference the main azure quick starts repo rather than my user's fork; do I need to change it back so it uses my forked copy?

bmoore-msft commented:
Looks ok, now - CI should update the branch in any url's so I think we're good to go...

CalCof commented:
Ok perfect, thanks a lot for all your help.

CalCof commented:
Do you know roughly how long it takes to get into the main repo? Also, if I wanted to submit two other ARM templates, should I do so in one pull or two separate pull requests?

CalCof commented:
I accidentally added some new files to this branch, but have removed them now. Do you know how long it will take for the pull request to complete?

CalCof commented:
Are you able to give any indication as to when this pull request will be merged?
